File: app/src/main/java/com/permissionnanny/simple/SimpleOperation.java
package com.permissionnanny.simple;

import android.support.v4.util.ArrayMap;
import com.permissionnanny.Operation;
import com.permissionnanny.ProxyFunction;
import com.permissionnanny.lib.request.RequestParams;

import java.util.Map;

/**
 *
 */
public class SimpleOperation extends Operation {
    public static final Map<String, SimpleOperation> operations = new ArrayMap<>();

    static {
        for (SimpleOperation operation : AccountOperation.operations) {
            operations.put(operation.mOpCode, operation);
        }
        for (SimpleOperation operation : LocationOperation.operations) {
            operations.put(operation.mOpCode, operation);
        }
        for (SimpleOperation operation : TelephonyOperation.operations) {
            operations.put(operation.mOpCode, operation);
        }
        for (SimpleOperation operation : SmsOperation.operations) {
            operations.put(operation.mOpCode, operation);
        }
        for (SimpleOperation operation : WifiOperation.operations) {
            operations.put(operation.mOpCode, operation);
        }
    }

    public static SimpleOperation getOperation(RequestParams params) {
        return operations.get(params.opCode);
    }

    public final String mOpCode;
    public final String mPermission;
    public final ProxyFunction mFunction;

    public SimpleOperation(String opCode,
                           String permission,
                           int protectionLevel,
                           int dialogTitle,
                           int minSdk,
                           ProxyFunction function) {
        super(dialogTitle, minSdk, protectionLevel);
        mOpCode = opCode;
        mPermission = permission;
        mFunction = function;
    }
}

File: app/src/main/kotlin/com/permissionnanny/simple/SimpleOperation.kt
package com.permissionnanny.simple

import android.content.Context
import android.os.Bundle
import android.support.v4.util.ArrayMap
import com.permissionnanny.Operation
import com.permissionnanny.lib.request.RequestParams

/**

 */
open class SimpleOperation(
        val opCode: String,
        val permission: String,
        protectionLevel: Int,
        dialogTitle: Int,
        minSdk: Int,
        val proxyFunction: ((Context, RequestParams, Bundle) -> Unit)?)
    : Operation(dialogTitle, minSdk, protectionLevel) {

    companion object {
        val operations: MutableMap<String, SimpleOperation> = ArrayMap()

        init {
//            for (operation in AccountOperation.operations) {
//                operations.put(operation.opCode, operation)
//            }
//            for (operation in LocationOperation.operations) {
//                operations.put(operation.opCode, operation)
//            }
//            for (operation in TelephonyOperation.operations) {
//                operations.put(operation.opCode, operation)
//            }
//            for (operation in SmsOperation.operations) {
//                operations.put(operation.opCode, operation)
//            }
//            for (operation in WifiOperation.operations) {
//                operations.put(operation.opCode, operation)
//            }
        }

        fun getOperation(params: RequestParams): SimpleOperation? {
            if(operations.isEmpty()){
                for (operation in AccountOperation.operations) {
                    operations.put(operation.opCode, operation)
                }
                for (operation in LocationOperation.operations) {
                    operations.put(operation.opCode, operation)
                }
                for (operation in TelephonyOperation.operations) {
                    operations.put(operation.opCode, operation)
                }
                for (operation in SmsOperation.operations) {
                    operations.put(operation.opCode, operation)
                }
                for (operation in WifiOperation.operations) {
                    operations.put(operation.opCode, operation)
                }
            }
            return operations[params.opCode]
        }
    }
}
